relay-event-schema/src/protocol/span/convert.rs

```diff
@@ -0,0 +1,231 @@
+    ContextInner, Contexts, DefaultContext, Event, ProfileContext, Span, TraceContext,
+};
+
+use relay_protocol::Annotated;
+use std::collections::BTreeMap;
+
+macro_rules! map_fields {
+    (
+        top-level:
+            $(span.$span_field:ident <=> event.$event_field:ident), *
+        ;
+        contexts:
+        $(
+            $ContextType:ident:
+                $(span.$foo:ident $(, $(span.$span_field_from_context:ident),+)? <=> context.$context_field:ident), *
+            ;
+        )*
+        ;
+        fixed_for_span:
+            $(span.$fixed_span_field:ident <= $fixed_span_value:expr), *
+        ;
+        fixed_for_event:
+            $($fixed_event_value:expr => span.$fixed_event_field:ident), *
+    ) => {
+        #[allow(clippy::needless_update)]
+        impl From<&Event> for Span {
+            fn from(event: &Event) -> Self {
+                Self {
+                    $(
+                        $span_field: event.$event_field.clone(),
+                    )*
+                    $(
+                        $(
+                            $foo: event.context::<$ContextType>()
+                                .map_or(None, |ctx|ctx.$context_field.value().cloned()).into(),
+                            $(
+                                $(
+                                    $span_field_from_context: event.context::<$ContextType>()
+                                    .map_or(None, |ctx|ctx.$context_field.value().cloned()).into(),
+                                )+
+                            )?
+                        )*
+                    )*
+                    $(
+                        $fixed_span_field: $fixed_span_value.into(),
+                    )*
+                    ..Default::default()
+                }
+            }
+        }
+
+        #[allow(clippy::needless_update)]
+        impl From<&Span> for Event {
+            fn from(span: &Span) -> Self {
+                Self {
+                    $(
+                        $event_field: span.$span_field.clone(),
+                    )*
+                    $(
+                        $fixed_event_field: $fixed_event_value.into(),
+                    )*
+                    contexts: Annotated::new(
+                        Contexts(
+                            BTreeMap::from([
+                                $(
+                                    (<$ContextType as DefaultContext>::default_key().into(), ContextInner($ContextType {
+                                        $(
+                                            $context_field: span.$foo.clone(),
+                                        )*
+                                        ..Default::default()
+                                    }.into_context()).into()),
+                                )*
+                            ]),
+                        )
+                    ),
+                    ..Default::default()
+                }
+            }
+        }
+    };
+}
```

Comment:
I'm considering moving the macro definition itself to a different module, just so users of the macro don't have to scroll down every time the want to update the macro invocation.